src/app/icd/client/CheckInHandler.cpp

```diff
@@ -0,0 +1,99 @@
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file defines objects for a CHIP ICD handler which handles unsolicited checkin messages.
+ *
+ */
+
+#include "CheckInHandler.h"
+#include "CheckInDelegate.h"
+
+#include <cinttypes>
+
+#include <lib/core/Global.h>
+#include <lib/support/CodeUtils.h>
+#include <messaging/Flags.h>
+#include <protocols/Protocols.h>
+
+#include <protocols/secure_channel/Constants.h>
+
+namespace chip {
+namespace app {
+
+static Global<CheckInMessageHandler> sCheckInMessageHandler;
+CheckInMessageHandler * CheckInMessageHandler::GetInstance()
+{
+    return &sCheckInMessageHandler.get();
+}
+
+CHIP_ERROR CheckInMessageHandler::Init(Messaging::ExchangeManager * exchangeManager, ICDClientStorage * clientStorage)
+{
+    VerifyOrReturnError(exchangeManager != nullptr, CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrReturnError(clientStorage != nullptr, CHIP_ERROR_INVALID_ARGUMENT);
+    mExchangeManager  = exchangeManager;
+    mICDClientStorage = static_cast<DefaultICDClientStorage *>(clientStorage);
+    ReturnErrorOnFailure(
+        exchangeManager->RegisterUnsolicitedMessageHandlerForType(Protocols::SecureChannel::MsgType::ICD_CheckIn, this));
+
+    return CHIP_NO_ERROR;
+}
+
+void CheckInMessageHandler::Shutdown()
+{
+    if (mExchangeManager)
+    {
+        mExchangeManager->UnregisterUnsolicitedMessageHandlerForType(Protocols::SecureChannel::MsgType::ICD_CheckIn);
+        mExchangeManager = nullptr;
+    }
+}
+CHIP_ERROR CheckInMessageHandler::OnUnsolicitedMessageReceived(const PayloadHeader & payloadHeader, ExchangeDelegate *& newDelegate)
+{
+    // Return error for wrong message type
+    VerifyOrReturnError(payloadHeader.HasMessageType(Protocols::SecureChannel::MsgType::ICD_CheckIn), CHIP_ERROR_INVALID_ARGUMENT);
+
+    newDelegate = this;
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR CheckInMessageHandler::OnMessageReceived(Messaging::ExchangeContext * ec, const PayloadHeader & payloadHeader,
+                                                    System::PacketBufferHandle && payload)
+{
+    VerifyOrReturnError(payloadHeader.HasMessageType(Protocols::SecureChannel::MsgType::ICD_CheckIn), CHIP_ERROR_INVALID_ARGUMENT);
+
+    ByteSpan payloadByteSpan{ payload->Start(), payload->DataLength() };
+    auto * iterator = mICDClientStorage->IterateICDClientInfo();
+    CHIP_ERROR err  = CHIP_NO_ERROR;
+    ICDClientInfo clientInfo;
+    while (iterator->Next(clientInfo))
+    {
+        err = mICDClientStorage->ProcessCheckInPayload(payloadByteSpan, clientInfo);
```

Comment:
clientInfo is output. You need move the clientInfo loop inside ProcessCheckInPayload.
Here, you just need
ByteSpan payloadByteSpan{ payload->Start(), payload->DataLength() };
VerifyOrReturnError(mICDClientStorage->ProcessCheckInPayload(payloadByteSpan, clientInfo));
return mpCheckInDelegate->OnCheckInComplete();

Reply:
Is it possible to pass the matched node id (or peer id)  to the OnCheckInComplete()? Or the app cannot know which device is active now?

Reply:
offline chat, we should pass clusterInfo instead of nodeId.